File: rxutil-rxjava/src/main/java/com/inqbarna/rxutil/paging/RxPaginatedList.java
/* 
 * Copyright 2014 InQBarna Kenkyuu Jo SL 
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); 
 * you may not use this file except in compliance with the License. 
 * You may obtain a copy of the License at 
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0 
 * 
 * Unless required by applicable law or agreed to in writing, software 
 * distributed under the License is distributed on an "AS IS" BASIS, 
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
 * See the License for the specific language governing permissions and 
 * limitations under the License. 
 */ 

package com.inqbarna.rxutil.paging;

import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;

import com.inqbarna.common.AdapterSyncList;
import com.inqbarna.common.paging.PaginatedAdapterDelegate;
import com.inqbarna.common.paging.PaginatedList;

import java.util.AbstractList;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Iterator;
import java.util.List;
import java.util.ListIterator;

import rx.Observable;
import rx.Subscriber;

/**
 * @author David García (david.garcia@inqbarna.com)
 * @version 1.0 16/9/16
 */

public class RxPaginatedList<U> extends Subscriber<List<? extends U>> implements PaginatedList<U> {
    private final Callbacks mCallbacks;
    private final RxPagingConfig mConfig;
    private List<U> mData;
    private boolean mCompleted;

    interface Callbacks extends RxPagingCallback {
        void onItemsAdded(int startPos, int size);
    }

    public static <T> PaginatedList<T> create(Observable<? extends List<? extends T>> stream, Callbacks callbacks, RxPagingConfig config) {
        return new RxPaginatedList<>(stream, callbacks, config);
    }

    private RxPaginatedList(Observable<? extends List<? extends U>> stream, Callbacks callbacks, RxPagingConfig config) {
        mCallbacks = callbacks;
        mConfig = config;
        mData = new ArrayList<>();
        mCompleted = false;
        stream.subscribe(this);
    }

    @Override
    public void onStart() {
        super.onStart();
        requestNext();
    }

    @Override
    public U get(int location) {
        return null == mData ? null : mData.get(location);
    }

    @Override
    public int size() {
        return null == mData ? 0 : mData.size();
    }

    @Override
    public boolean hasMorePages() {
        return !mCompleted;
    }

    @Override
    public void requestNext() {
        if (!mCompleted) {
            request(1);
            return;
        }
        throw new IllegalStateException("You requested data after completed!");
    }

    @Override
    public void appendPageItems(Collection<? extends U> items, boolean last) {
        throw new UnsupportedOperationException("This is not supported, only internal data flow will be accepted");
    }

    @Override
    public void clear(@Nullable PaginatedAdapterDelegate.ItemRemovedCallback<U> callback) {
        if (null == callback) {
            mData.clear();
        } else {
            final Iterator<U> iterator = mData.iterator();
            while (iterator.hasNext()) {
                callback.onItemRemoved(iterator.next());
                iterator.remove();
            }
        }
        mCompleted = true;
        mCallbacks.onCompleted();
        unsubscribe();
    }

    @Override
    public void onCompleted() {
        mCompleted = true;
        mCallbacks.onCompleted();
    }

    @Override
    public void onError(Throwable e) {
        mCompleted = true;
        mCallbacks.onError(e);
    }

    @Override
    public void onNext(List<? extends U> us) {
        final int startSize = mData.size();
        mData.addAll(us);
        mCallbacks.onItemsAdded(startSize, us.size());
    }

    @Override
    public List<U> editableList(@Nullable RecyclerView.Adapter callbackAdapter) {
        return new AdapterSyncList<>(mData, callbackAdapter);
    }

}

File: rxutil-rxjava2/src/main/java/com/inqbarna/rxutil/paging/RxPaginatedList.kt
package com.inqbarna.rxutil.paging

import androidx.recyclerview.widget.RecyclerView
import com.inqbarna.common.AdapterSyncList
import com.inqbarna.common.paging.PaginatedAdapterDelegate
import com.inqbarna.common.paging.PaginatedList
import io.reactivex.Flowable
import io.reactivex.subscribers.DisposableSubscriber

/**
 * @author David García <david.garcia></david.garcia>@inqbarna.com>
 * @version 1.0 16/9/16
 */

class RxPaginatedList<U> private constructor(
        stream: Flowable<out List<U>>,
        private val callbacks: Callbacks,
        private val rxConfig: RxPagingConfig
) : DisposableSubscriber<List<U>>(), PaginatedList<U> {

    private val data: MutableList<U> = mutableListOf()
    private var completed: Boolean = false

    internal interface Callbacks : RxPagingCallback {
        fun onItemsAdded(startPos: Int, size: Int)
    }

    init {
        completed = false
        stream.subscribe(this)
    }

    override fun onStart() {
        requestNext()
    }

    override fun get(location: Int): U? {
        return data.getOrNull(location)
    }

    override fun size(): Int {
        return data.size
    }

    override fun hasMorePages(): Boolean = !completed

    override fun requestNext() {
        if (!isDisposed) {
            request(1)
            return
        }
        throw IllegalStateException("You requested data after completed!")
    }

    override fun appendPageItems(items: Collection<U>, last: Boolean) {
        throw UnsupportedOperationException("This is not supported, only internal data flow will be accepted")
    }

    override fun clear(itemRemovedCallback: PaginatedAdapterDelegate.ItemRemovedCallback<U>?) {
        if (null != itemRemovedCallback) {
            val iterator = data.iterator()
            while (iterator.hasNext()) {
                itemRemovedCallback.onItemRemoved(iterator.next())
                iterator.remove()
            }
        } else {
            data.clear()
        }
        completed = true
        callbacks.onCompleted()
        dispose()
    }

    override fun onComplete() {
        completed = true
        callbacks.onCompleted()
        dispose()
    }

    override fun onError(e: Throwable) {
        completed = true
        callbacks.onError(e)
        dispose()
    }

    override fun onNext(us: List<U>) {
        val startSize = data.size
        data.addAll(us)
        callbacks.onItemsAdded(startSize, us.size)
    }

    override fun editableList(callbackAdapter: RecyclerView.Adapter<*>?): List<U> {
        return AdapterSyncList(data, callbackAdapter)
    }

    companion object {
        internal fun <T> create(stream: Flowable<out List<T>>, callbacks: Callbacks, config: RxPagingConfig): PaginatedList<T> {
            return RxPaginatedList(stream, callbacks, config)
        }
    }

}
